pandas/core/internals/blocks.py

```diff
@@ -974,10 +974,19 @@ def setitem(self, indexer, value) -> Block:
                     # checking lib.is_scalar here fails on
                     #  test_iloc_setitem_custom_object
                     casted = setitem_datetimelike_compat(values, len(vi), casted)
+
+            if using_cow and self.refs.has_reference():
+                values = values.copy()
+                self = type(self)(
```

Comment:
It seems we have a mix of using `type(self)(..)` and `self.make_block_same_class(..)`. The latter avoids having to pass mgr_locs and ndim)

Reply:
Good idea, switched.